### PR TITLE
FSET-1711: Progress to FSB or Job offer from assessment_centre_passed

### PR DESCRIPTION
**Pull request by henricook**

### CODE REVIEW

File: app/repositories/application/GeneralApplicationRepository.scala [hunk 1]
```diff
@@ -225,21 +225,25 @@ class GeneralApplicationMongoRepository(
         val applicationStatus = document.getAs[ApplicationStatus]("applicationStatus").get
         val applicationRoute = document.getAs[ApplicationRoute]("applicationRoute").getOrElse(ApplicationRoute.Faststream)
         val progressStatusTimeStampDoc = document.getAs[BSONDocument]("progress-status-timestamp")
-        val progressStatusTimeStamp = progressStatusTimeStampDoc.flatMap { timestamps =>
+        val latestProgressStatus = progressStatusTimeStampDoc.flatMap { timestamps =>
           val relevantProgressStatuses = timestamps.elements.filter(_._1.startsWith(applicationStatus))
+          val latestRelevantProgressStatus = relevantProgressStatuses.maxBy(element => timestamps.getAs[DateTime](element._1).get)
+          Try(ProgressStatuses.nameToProgressStatus(latestRelevantProgressStatus._1)) match {
+            case Success(progressStatus) => Some(progressStatus)
```
Comment (vglushak) [inline app/repositories/application/GeneralApplicationRepository.scala]: You can use `.toOption` function on `Try()`

File: app/services/assessmentcentre/AssessmentCentreToFsbOrOfferProgressionService.scala [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    fsbRepo.nextApplicationForFsbOrJobOfferProgression(batchSize)
+  }
+
+  def progressApplicationsToFsbOrJobOffer(applications: Seq[ApplicationForProgression])
+  : Future[SerialUpdateResult[ApplicationForProgression]] = {
+
+    def maybeProgressToFsbOrJobOffer(application: ApplicationForProgression, firstResidual: SchemeEvaluationResult): Future[Unit] = {
+      if (firstResidual.result == Green.toString && fsbRequiredSchemeIds.contains(firstResidual.schemeId)) {
+        fsbRepo.progressToFsb(application)
+      } else if (firstResidual.result == Green.toString && !fsbRequiredSchemeIds.contains(firstResidual.schemeId)) {
```
Comment (vglushak) [inline app/services/assessmentcentre/AssessmentCentreToFsbOrOfferProgressionService.scala]: second condition looks excessive.
I would do it as pattern matching on `result` with guard condition on `fsbRequiredSchemeIds` check.

File: app/services/assessmentcentre/AssessmentCentreToFsbOrOfferProgressionService.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      } else {
+        Future.successful(())
+      }
+    }
+
+    val updates = FutureEx.traverseSerial(applications) { application =>
+      FutureEx.futureToEither(application,
+        for {
+          currentSchemeStatus <- applicationRepo.getCurrentSchemeStatus(application.applicationId)
+          firstResidual = firstResidualPreference(currentSchemeStatus).get
```
Comment (vglushak) [inline app/services/assessmentcentre/AssessmentCentreToFsbOrOfferProgressionService.scala]: what if candidate has failed everything - `None.get` ?

File: app/model/ProgressStatuses.scala [hunk 4]
```diff
@@ -121,6 +121,7 @@ object ProgressStatuses {
   case object ASSESSMENT_CENTRE_AWAITING_RE_EVALUATION extends ProgressStatus(ApplicationStatus.ASSESSMENT_CENTRE)
   case object ASSESSMENT_CENTRE_PASSED extends ProgressStatus(ApplicationStatus.ASSESSMENT_CENTRE)
   case object ASSESSMENT_CENTRE_FAILED extends ProgressStatus(ApplicationStatus.ASSESSMENT_CENTRE)
+  case object ASSESSMENT_CENTRE_FAILED_NOTIFIED extends ProgressStatus(ApplicationStatus.ASSESSMENT_CENTRE)
```
Comment (vglushak) [inline app/model/ProgressStatuses.scala]: Thanks for adding this for me.

I know don't look a gift-horse in the mouth, but maybe you can also add `FSB_FAILED_NOTIFIED` 😳

File: app/repositories/application/GeneralApplicationRepository.scala (hunk 1 above)
Comment (henricook) [reply app/repositories/application/GeneralApplicationRepository.scala]: :+1:

File: app/services/assessmentcentre/AssessmentCentreToFsbOrOfferProgressionService.scala (hunk 3 above)
Comment (henricook) [reply app/services/assessmentcentre/AssessmentCentreToFsbOrOfferProgressionService.scala]: Shouldn't be possible at this stage, they shouldn't be getting passed to this function without an eligible scheme. In that case this .get fails and it comes back as a failure in SerialUpdateResult

File: app/model/ProgressStatuses.scala (hunk 4 above)
Comment (henricook) [reply app/model/ProgressStatuses.scala]: Go on then :roll_eyes:

**Pull request merged**